File: src/p3xxx/p3117.py
from leetcode_prelude import *


# Problem 3117
class Solution:
    def minimumValueSum(self, nums: List[int], andValues: List[int]) -> int:
        and_len = len(andValues)

        ones = 2**30 - 1
        dp0 = {}
        dp0[(ones, 0)] = 0

        dp1 = {}

        for number in nums:
            for (bits, and_idx), value in dp0.items():
                if and_idx >= and_len:
                    continue

                bits &= number
                if bits >= andValues[and_idx]:
                    key = (bits, and_idx)

                    new_value = dp1.get(key)
                    if new_value is None:
                        new_value = value
                    else:
                        new_value = min(value, new_value)

                    dp1[key] = new_value

                if bits == andValues[and_idx]:
                    key = (ones, and_idx + 1)
                    new_value = dp1.get(key)
                    if new_value is None:
                        new_value = value + number
                    else:
                        new_value = min(new_value, value + number)
                    dp1[key] = new_value

            dp0, dp1 = dp1, dp0
            dp1.clear()

        return dp0.get((ones, and_len), -1)

File: src/p3xxx/p3117.rs
mod leetcode_prelude;

use leetcode_prelude::*;

pub fn main() {}

// hello world !!!!

extern crate rand;

// Problem 3117
impl Solution {
    pub fn minimum_value_sum2(nums: Vec<i32>, and_values: Vec<i32>) -> i32 {
        /*
         * HashMap is slow.
         * Observe the pattern of the hash map
         * Given an and_idx, there are at most two bits.
         * We may convert the hash map into array-like things, but it is more complex
         * due to duplicated code.
         */

        use std::collections::HashMap;

        let mut dp0 = HashMap::<(i32, usize), i32>::new();
        let mut dp1 = dp0.clone();

        dp0.insert((i32::MAX, 0), 0);

        for number in nums.iter().copied() {
            for ((bits, and_idx), value) in dp0.drain() {
                if and_idx >= and_values.len() {
                    continue;
                }

                let bits = bits & number;

                if bits >= and_values[and_idx] {
                    dp1.entry((bits, and_idx))
                        .and_modify(|sum| *sum = value.min(*sum))
                        .or_insert(value);
                }

                if bits == and_values[and_idx] {
                    let and_idx = and_idx + 1;
                    let value = value + number;
                    dp1.entry((i32::MAX, and_idx))
                        .and_modify(|sum| *sum = value.min(*sum))
                        .or_insert(value);
                }
            }

            std::mem::swap(&mut dp0, &mut dp1);
            println!("{:?}", dp0);
        }

        dp0.get(&(i32::MAX, and_values.len()))
            .copied()
            .unwrap_or(-1)
    }
}
